tests/cycle_fixpoint.rs

```diff
@@ -0,0 +1,207 @@
+/// Test case for fixpoint iteration cycle resolution.
+use salsa::{CycleRecoveryAction, Database as Db, Setter};
+use std::collections::BTreeSet;
+use std::iter::IntoIterator;
+
+/// A Use of a symbol.
+#[salsa::input]
+struct Use {
+    reaching_definitions: Vec<Definition>,
+}
+
+#[salsa::input]
+struct Literal {
+    value: usize,
+}
+
+/// A Definition of a symbol, either of the form `base + increment` or `0 + increment`.
+#[salsa::input]
+struct Definition {
+    base: Option<Use>,
+    increment: Literal,
+}
+
+#[derive(Eq, PartialEq, Clone, Debug)]
+enum Type {
+    Bottom,
+    Values(Box<[usize]>),
+    Top,
+}
+
+impl Type {
+    fn join(tys: impl IntoIterator<Item = Type>) -> Type {
+        let mut result = Type::Bottom;
+        for ty in tys.into_iter() {
+            result = match (result, ty) {
+                (result, Type::Bottom) => result,
+                (_, Type::Top) => Type::Top,
+                (Type::Top, _) => Type::Top,
+                (Type::Bottom, ty) => ty,
+                (Type::Values(a_ints), Type::Values(b_ints)) => {
+                    let mut set = BTreeSet::new();
+                    set.extend(a_ints);
+                    set.extend(b_ints);
+                    Type::Values(set.into_iter().collect())
+                }
+            }
+        }
+        result
+    }
+}
+
+#[salsa::tracked(cycle_fn=cycle_recover, cycle_initial=cycle_initial)]
```

Comment:
Yes, I thought about this as well. It's not clear to me what is actually the better UX. Implementing a trait is somewhat more boilerplate, and in practice there's not much difference in the experience if you fail to implement one of the methods. Either way the compiler catches the error for you, because its an error in macro expansion if you don't provide both `cycle_fn` and `cycle_initial`, with a trait it would be a compiler error that you didn't fully implement the trait.

Reply:
I guess with a trait your IDE might give you the right signatures of the methods for free, which is kind of nice...

Reply:
I normally don't like traits, but I feel like a trait would be ideal preferable because the annotation seems pretty darn wordy.

Reply:
Ok, going to still wait for Niko's feedback on this point before updating, but the trait idea makes sense to me.

Reply:
Actually, I'm okay with a non-trait based approach. rust-analyzer uses a `#[salsa::cycle(path::to::function)]`-style annotation, so I think to ease the transition, avoiding a trait would be nice.

Reply:
>  rust-analyzer uses a #[salsa::cycle(path::to::function)]-style annotation, so I think to ease the transition, avoiding a trait would be nice.

Is cycle handling common in ra? I do see how it reduces the diff size because it isn't necessary to make the function a trait-method but you would still have to change every use because you now have to specify both the cycle and cycle initial functions.

Reply:
Added a TODO comment in the code for this question as well.

Reply:
its common but not common enough that it would matter (there are ~11 cycle handlers and the majority of them return a dummy result)

Reply:
One theoreitcal nice thing about taking the functions here directly would be that we could allow supporting closure expressions in that position inline instead of having to create a whole function (as cycles that are not going to be fixpoint are likely not gonna have complex functions here), a trait would be quite wordy imo.